File: AI_lab/bfs.py
#bfs: Traverse a graph in breadth first manner.
#Time complexity: O(V+E)
#Space complexity: O(V)

Graph = {
    '5' : ['3','7'],
    '3' : ['2', '4'],
    '7' : ['8'],
    '2' : [],
    '4' : ['8'],
    '8' : []
}

#function to implement bfs
def bfs(source):
    #to avoid duplication of nodes while traversal
    visited = []

    #to store the order for printing in bfs, fifo
    queue = []

    #add source to visited and queue
    queue.append(source)
    visited.append(source)

    #while queue is not empty, pop the first element, print the element and repeat same procedure for unvisited neighbours
    while(queue):
        v = queue.pop(0)
        print(v, end = " ")
        
        #check for the neighbours of popped vertex
        for nbr in Graph[v]:
            if nbr not in visited:
                visited.append(nbr)
                queue.append(nbr)
                
bfs('5')
#numVertices = input("Enter the number of vertices in graph:")

File: Graph/Course/BFS.cpp
#include<bits/stdc++.h>
using namespace std;

class Graph {
    int V;
    list<int> *l;
public:
    Graph(int v){
        V = v;
        l = new list<int>[V];
    }

    void addEdge(int i, int j, bool undir=true){
        l[i].push_back(j);
        if(undir){
            l[j].push_back(i);
        }
    }

    void printGraph() {
        for(int i=0; i<V; i++){
            cout << i << "-->";
            for(auto x: l[i]){
                cout << x << " ";
            }
            cout << endl;
        }
    }

    void bfs(int source) {
        queue<int> q;
        q.push(source);
        int visited[V] = {0};
        visited[source] = 1;
        //cout << "Okay";
        while(!q.empty()){
            int top = q.front();
            q.pop();
            for(auto nbr: l[top]){
                if(!visited[nbr]){
                    q.push(nbr);
                    visited[nbr] = 1;
                }
            }
            //cout << "Okat";
            cout << top << " ";
        }
        //cout << "Ok";
    }
};

int main() {
    Graph g(7);
    g.addEdge(0, 1);
    g.addEdge(1, 2);
    g.addEdge(2, 3);
    g.addEdge(3, 5);
    g.addEdge(5, 6);
    g.addEdge(4, 5);
    g.addEdge(0, 4);
    g.addEdge(3, 4);

    g.printGraph();
    g.bfs(1);
    return 0;
}